<a href="https://colab.research.google.com/github/201921027/graduate_project_git/blob/main/gp_data_processing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

In [3]:
csv_data = pd.read_csv('/content/drive/MyDrive/diary_sentiment_analysis/NIKL_NP.csv')

In [4]:
csv_data.head()

Unnamed: 0         file_id anno_level   category    title author_id  \
0           0  WDRW1900100013         원시  비출판물 > 수필  아내의 생일상    P00013   
1           1  WDRW1900100013         원시  비출판물 > 수필  아내의 생일상    P00013   
2           2  WDRW1900100013         원시  비출판물 > 수필  아내의 생일상    P00013   
3           3  WDRW1900100013         원시  비출판물 > 수필  아내의 생일상    P00013   
4           4  WDRW1900100013         원시  비출판물 > 수필  아내의 생일상    P00013   

   author_age author_occupation author_sex author_submission  \
0          45           직장인/전문직          M               온라인   
1          45           직장인/전문직          M               온라인   
2          45           직장인/전문직          M               온라인   
3          45           직장인/전문직          M               온라인   
4          45           직장인/전문직          M               온라인   

  author_handwriting  text_date text_subclass  \
0                 No   20200000      null_생일상   
1                 No   20200000      null_생일상   
2                 No   20200000      null_생일상   
3                 No   20200000      null_생일상   
4                 No   20200000      null_생일상   

                                            sentence  
0  어제는 아내의 생일이었다. 생일을 맞이하여 아침을 준비하겠다고 오전 8시 30분부터...  
1  주된 메뉴는 스테이크와 낙지볶음, 미역국, 잡채, 소야 등이었다. 스테이크는 자주하...  
2  그런데 상상도 못한 일이 벌이지고 말았다. 보통 시즈닝이 되지 않은 원육을 사서 스...  
3  앞면을 센불에 1분을 굽고 뒤집는 순간 방부제가 함께 구어진 것을 알았다. 아내의 ...  
4  어처구니 없는 상황이 발생한 것이다. 방부제가 센불에 녹아서 그런지 물처럼 흘러내렸...

In [5]:
print(csv_data['category'].nunique())
print(csv_data['category'].unique())

518
['비출판물 > 수필' '비출판물 > 감상문' '비출판물 > 일기' '비출판물 > 기타 > 칼럼' '비출판물 > 편지글'
 '비출판물 > 소설(동화)' '비출판물 > 시(동시)' '비출판물 > 수필 > 나의 삶'
 '비출판물 > 기타 > 전문 개발 프론트엔드의 개발서' '비출판물 > 기타 > 자기소개서' '비출판물 > 기타 > 수필'
 '비출판물 > 기타 > 시나리오 기획안' '비출판물 > 기타 > 사회비평' '비출판물 > 기타 > 숲에서'
 '비출판물 > 기타 > 자유주제' '비출판물 > 기타 > 아이돌실물후기' '비출판물 > 감상문 > 독서감상문'
 '비출판물 > 기타 > 소설' '비출판물 > 기타 > 주의사항' '비출판물 > 일기 > 2020년 2월 5일의 일기'
 '비출판물 > 기타 > 비평문' '비출판물 > 수필 > 주제있는 수필' '비출판물 > 기타 > 종합소설'
 '비출판물 > 감상문 > 사진 전시 감상문' '비출판물 > 기타 > 음악사 인물 캐릭터 창작'
 '비출판물 > 기타 > 인공지능으로 대체되어야 할 직업' '비출판물 > 기타 > 창작소설' '비출판물 > 기타 > 에세이'
 '비출판물 > 기타 > 여행기' '비출판물 > 소설(동화) > 인물 회상 조각글' '비출판물 > 기타 > 분석글'
 '비출판물 > 기타 > 건의문' '비출판물 > 기타 > 음식' '비출판물 > 기타 > 전시회 참가 후 느낀점'
 '비출판물 > 기타 > 관찰기록' '비출판물 > 기타 > 상명대학교 면접 예상질답' '비출판물 > 기타 > 현장견학 보고서'
 '비출판물 > 기타 > 저널' '비출판물 > 기타 > 기행문' '비출판물 > 기타 > 대입 자기소개서'
 '비출판물 > 기타 > 조사 보고서' '비출판물 > 기타 > 자기소개' '비출판물 > 기타 > 보고서'
 '비출판물 > 기타 > 자소서' '비출판물 > 소설(동화) > 독백' '비출판물 > 기타 > 묘사의 습작'
 '비출판물 > 기타 > 백현에 대한 생각' '비출판물 > 기타 > 코스프레이야기' '비출판물 > 

In [6]:
filter_category= ['비출판물 > 일기']

dfs = []
for i in range(len(filter_category)):
    temp = csv_data.loc[csv_data['category'] == filter_category[i]]
    dfs.append(temp)
filter_csv = pd.concat(dfs, ignore_index=True)
filter_csv.head()

Unnamed: 0         file_id anno_level   category   title author_id  \
0          10  WDRW1900100023         원시  비출판물 > 일기      졸업    P00023   
1          26  WDRW1900100037         원시  비출판물 > 일기  58년 개띠    P00037   
2          27  WDRW1900100037         원시  비출판물 > 일기  58년 개띠    P00037   
3          28  WDRW1900100037         원시  비출판물 > 일기  58년 개띠    P00037   
4          29  WDRW1900100037         원시  비출판물 > 일기  58년 개띠    P00037   

   author_age author_occupation author_sex author_submission  \
0          24          대학생/대학원생          F               온라인   
1          61              전업주부          M               온라인   
2          61              전업주부          M               온라인   
3          61              전업주부          M               온라인   
4          61              전업주부          M               온라인   

  author_handwriting  text_date text_subclass  \
0                 No   20200000            학교   
1                 No   20200000      null_노년기   
2                 No   20200000      null_노년기   
3                 No   20200000      null_노년기   
4                 No   20200000      null_노년기   

                                            sentence  
0  졸업식이 취소되어서 너무 아쉽다. 취소 이유는 중국 우한에서 처음 발병한 신종 코로...  
1  살 다 보니 어느 새 현역에서 은퇴하여 손주를 보며 즐거워하는 나이가 되었다. 지난...  
2  그러다가 어떤 계기가 되어 한 10년 열심히 해서 집 한 채 마련하여 평범한 삶을 ...  
3  사람이 태어나 살아가면서 열심히 일하면 일한 만큼의 대가를 얻어 의식주에 대한 고민...  
4  우리나라의 결혼식 문화는 한 번이라도 안면이 있는 사람은 본인 혹은 자녀의 결혼에 ...

In [7]:
print(len(filter_csv))

22485


In [8]:
# groupby 후 sentence를 합치기 (띄어쓰기로 연결)
df= (
    filter_csv.groupby(['category','title', 'author_id', 'text_date'], as_index=False)
      .agg({'sentence': ' '.join})
      .rename(columns={'sentence': 'text'})
)

# 결과 확인
print(df.head())

    category              title author_id  text_date  \
0  비출판물 > 일기    5학년을 마치며....       P12574   20200219   
1  비출판물 > 일기     <name1>의 깜짝 등장    P12194   20180120   
2  비출판물 > 일기       <name2>집에 초대    P12490   20190803   
3  비출판물 > 일기              Happy    P11531   20190525   
4  비출판물 > 일기                 TV    P11691   20190602   

                                                text  
0  5학년을 마치며....                                  ...  
1  2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...  
2  <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...  
3  다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...  
4  앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...  


In [9]:
df.head()

category              title author_id  text_date  \
0  비출판물 > 일기    5학년을 마치며....       P12574   20200219   
1  비출판물 > 일기     <name1>의 깜짝 등장    P12194   20180120   
2  비출판물 > 일기       <name2>집에 초대    P12490   20190803   
3  비출판물 > 일기              Happy    P11531   20190525   
4  비출판물 > 일기                 TV    P11691   20190602   

                                                text  
0  5학년을 마치며....                                  ...  
1  2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...  
2  <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...  
3  다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...  
4  앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...

In [10]:
print(len(df))

4180


In [11]:
df.isnull().sum()

category     0
title        0
author_id    0
text_date    0
text         0
dtype: int64

In [12]:
# text 컬럼의 글자 수 계산
df['text_length'] = df['text'].str.len()

df

category              title author_id  text_date  \
0     비출판물 > 일기    5학년을 마치며....       P12574   20200219   
1     비출판물 > 일기     <name1>의 깜짝 등장    P12194   20180120   
2     비출판물 > 일기       <name2>집에 초대    P12490   20190803   
3     비출판물 > 일기              Happy    P11531   20190525   
4     비출판물 > 일기                 TV    P11691   20190602   
...         ...                ...       ...        ...   
4175  비출판물 > 일기                힘들다    P10532   20200000   
4176  비출판물 > 일기         힘들면서 좋았던 날    P10065   20200000   
4177  비출판물 > 일기            힘들었던 날.    P02733   20200000   
4178  비출판물 > 일기       힘은 솟아나지 않는다.    P03120   20200000   
4179  비출판물 > 일기        힘을 내라고 말해줄래    P03963   20200000   

                                                   text  text_length  
0     5학년을 마치며....                                  ...          370  
1     2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...          273  
2     <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...          167  
3     다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...          141  
4     앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...          117  
...                                                 ...          ...  
4175  요즈음 엄마가 회사 일 때문에 바쁘셔서 밥을 못 챙겨 주신다. 나는 밥과 반찬을 잘...          419  
4176  <힘들면서 좋았던 날> 오늘 아침 8시에 일어났다. 조금 피곤했지만 또 잠들기는 싫...          575  
4177  오늘은 감정적으로 매우 힘든 날이었다. 우울하고 지치고... 나갈 곳은 없고 하고싶...         1040  
4178  힘은 솟아나지 않는다. 힘이란 연속되면서 이어진다. 관성과 같다. 나는 언제나 솟구...         1259  
4179  같이 게임하는 친구가 지칠때마다 힘들어도.. 힘내자! 이러는데 그 말이 너무 웃김 ...          332  

[4180 rows x 6 columns]

In [13]:
bins = [0, 50, 100, 200, 300, 400, 500, 600, 700, 800, 900, float('inf')]
labels = ['0-50', '51-100', '101-200', '201-300', '301-400', '401-500', '501-600', '601-700', '701-800', '801-900', '901+']

df['length_bin'] = pd.cut(df['text_length'], bins=bins, labels=labels, right=True)
count_by_bin = df['length_bin'].value_counts().sort_index()

print(count_by_bin)


length_bin
0-50         34
51-100      208
101-200     971
201-300     626
301-400     471
401-500     316
501-600     201
601-700     139
701-800      86
801-900      49
901+       1079
Name: count, dtype: int64


In [15]:
train_df = df.drop(columns=['author_id', 'text_date','text_length', 'length_bin'])

# CSV 파일로 저장
train_df.to_csv('/content/drive/MyDrive/diary_sentiment_analysis/train_df.csv', index=False)

# Training Dataset 감정 매핑하기: 기존 감정 분석 사전학습 모델 기반

## 다중 감정 분석 모델: https://huggingface.co/searle-j/kote_for_easygoing_people

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!pip install kss

In [ ]:
!pip install --upgrade transformers

In [ ]:
#from transformers import AutoModelForSequenceClassification

#model_name = "searle-j/kote_for_easygoing_people"
#model = AutoModelForSequenceClassification.from_pretrained(m#odel_name)
#id2label = model.config.id2label
#print(id2label)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/4.10k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/498M [00:00<?, ?B/s]

{0: '불평/불만', 1: '환영/호의', 2: '감동/감탄', 3: '지긋지긋', 4: '고마움', 5: '슬픔', 6: '화남/분노', 7: '존경', 8: '기대감', 9: '우쭐댐/무시함', 10: '안타까움/실망', 11: '비장함', 12: '의심/불신', 13: '뿌듯함', 14: '편안/쾌적', 15: '신기함/관심', 16: '아껴주는', 17: '부끄러움', 18: '공포/무서움', 19: '절망', 20: '한심함', 21: '역겨움/징그러움', 22: '짜증', 23: '어이없음', 24: '없음', 25: '패배/자기혐오', 26: '귀찮음', 27: '힘듦/지침', 28: '즐거움/신남', 29: '깨달음', 30: '죄책감', 31: '증오/혐오', 32: '흐뭇함(귀여움/예쁨)', 33: '당황/난처', 34: '경악', 35: '부담/안_내킴', 36: '서러움', 37: '재미없음', 38: '불쌍함/연민', 39: '놀람', 40: '행복', 41: '불안/걱정', 42: '기쁨', 43: '안심/신뢰'}


In [ ]:
import pandas as pd
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TextClassificationPipeline
from kss import split_sentences
from tqdm import tqdm

# KOTE 43개 감정 → 13개 감정(형용사형+중립) 매핑 #"외로운' 없음.
kote_to_13 = kote_to_13 = {
    '불평/불만': '분노한',
    '환영/호의': '감사한',
    '감동/감탄': '감사한',
    '지긋지긋': '분노한',
    '고마움': '감사한',
    '슬픔': '슬픈',
    '화남/분노': '분노한',
    '존경': '감사한',
    '기대감': '흥분한',
    '우쭐댐/무시함': '자신하는',
    '안타까움/실망': '슬픈',
    '비장함': '자신하는',
    '의심/불신': '불안한',
    '뿌듯함': '만족스러운',
    '편안/쾌적': '편안한',
    '신기함/관심': '흥분한',
    '아껴주는': '감사한',
    '부끄러움': '당황한',
    '공포/무서움': '불안한',
    '절망': '우울한',
    '한심함': '우울한',
    '역겨움/징그러움': '분노한',
    '짜증': '분노한',
    '어이없음': '분노한',
    '없음': '그저 그런',
    '패배/자기혐오': '우울한',
    '귀찮음': '스트레스 받는',
    '힘듦/지침': '스트레스 받는',
    '즐거움/신남': '기쁜',
    '깨달음': '만족스러운',
    '죄책감': '슬픈',
    '증오/혐오': '분노한',
    '흐뭇함(귀여움/예쁨)': '기쁜',
    '당황/난처': '당황한',
    '경악': '당황한',
    '부담/안_내킴': '스트레스 받는',
    '서러움': '슬픈',
    '재미없음': '우울한',
    '불쌍함/연민': '슬픈',
    '놀람': '흥분한',
    '행복': '기쁜',
    '불안/걱정': '불안한',
    '기쁨': '기쁜',
    '안심/신뢰': '편안한'
    # '외로운'에 직접적으로 해당하는 라벨은 KOTE 원본에 없음
}


selected_emotions = [
    "기쁜", "감사한", "편안한", "만족스러운", "흥분한", "자신하는",
    "분노한", "슬픈", "우울한", "불안한", "스트레스 받는", "당황한", "그저 그런"
]

def chunk_text_with_kss(text, tokenizer, max_tokens=512):
    sentences = split_sentences(str(text))
    chunks = []
    current_chunk = []
    current_length = 0
    for sentence in sentences:
        encoded = tokenizer.encode(sentence, add_special_tokens=True)
        tokenized_len = len(encoded)
        if tokenized_len > max_tokens:
            # 단어 단위로 강제 분할
            words = sentence.split()
            chunk = []
            chunk_len = 0
            for word in words:
                word_encoded = tokenizer.encode(word, add_special_tokens=False)
                word_len = len(word_encoded)
                if chunk_len + word_len > max_tokens:
                    chunks.append(' '.join(chunk))
                    chunk = []
                    chunk_len = 0
                chunk.append(word)
                chunk_len += word_len
            if chunk:
                chunks.append(' '.join(chunk))
            current_length = 0
            current_chunk = []
            continue
        if current_length + tokenized_len <= max_tokens:
            current_chunk.append(sentence)
            current_length += tokenized_len
        else:
            chunks.append(' '.join(current_chunk))
            current_chunk = [sentence]
            current_length = tokenized_len
    if current_chunk:
        chunks.append(' '.join(current_chunk))
    return chunks

def kote_predict(text, pipe, kote_to_13, threshold=0.4, top_n=3, final_threshold=0.6):
    tokenizer = pipe.tokenizer
    chunks = chunk_text_with_kss(text, tokenizer)
    chunk_results = []
    all_emotions = []
    for chunk in chunks:
        outputs = pipe(chunk)
        mapped_scores = []
        for output_list in outputs:
            for out in output_list:
                mapped = kote_to_13.get(out['label'])
                if mapped in selected_emotions and out['score'] >= threshold:
                    mapped_scores.append((mapped, out['score']))
        mapped_scores = sorted(mapped_scores, key=lambda x: x[1], reverse=True)
        seen = set()
        top_emotions = []
        for emotion, score in mapped_scores:
            if emotion not in seen:
                top_emotions.append((emotion, score))
                seen.add(emotion)
            if len(top_emotions) >= top_n:
                break
        chunk_results.append(dict(top_emotions))
        all_emotions.extend([e for e, _ in top_emotions])
    total_emotions = {}
    for cr in chunk_results:
        for k, v in cr.items():
            if k not in total_emotions or v > total_emotions[k]:
                total_emotions[k] = v
    if not total_emotions or all(v < threshold for v in total_emotions.values()):
        emotions_prob = {'그저 그런': 1.0}
        exist_emotions = ['그저 그런']
    else:
        emotions_prob = total_emotions
        # 확률 내림차순 정렬 후 상위 3개만
        sorted_emotions = sorted(
            [(emotion, prob) for emotion, prob in emotions_prob.items() if prob >= final_threshold],
            key=lambda x: x[1], reverse=True
        )
        exist_emotions = [emotion for emotion, prob in sorted_emotions[:3]]
        if not exist_emotions:
            exist_emotions = ['그저 그런']
    return emotions_prob, exist_emotions



# ------------------- CSV 파일 처리 -------------------

input_csv = "/content/drive/MyDrive/diary_sentiment_analysis/train_df.csv"
output_csv = "/content/drive/MyDrive/diary_sentiment_analysis/train_df_emotions_1.csv"

df = pd.read_csv(input_csv)

model_name = "searle-j/kote_for_easygoing_people"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)
pipe = TextClassificationPipeline(
    model=model,
    tokenizer=tokenizer,
    function_to_apply='sigmoid',
    return_all_scores=True,
    device=0,  # GPU 사용
    truncation=True,  # 추가
    max_length=512    # 추가
)

emotions_prob_list = []
exist_emotions_list = []

for text in tqdm(df['text'], desc="감정 분석 중"):
    try:
        emotions_prob, exist_emotions = kote_predict(
            text, pipe, kote_to_13, threshold=0.4, top_n=3, final_threshold=0.6
        )
    except Exception as e:
        print(f"Error: {str(e)}")
        emotions_prob = {'error': 1.0}
        exist_emotions = ['error']
    emotions_prob_list.append(str(emotions_prob))
    exist_emotions_list.append(str(exist_emotions))

df['emotions_prob'] = emotions_prob_list
df['exist_emotions'] = exist_emotions_list

df.to_csv(output_csv, index=False)
print(f"분석 결과가 저장되었습니다: {output_csv}")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/545 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/396k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/788k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Device set to use cuda:0
/usr/local/lib/python3.11/dist-packages/transformers/pipelines/text_classification.py:106: UserWarning: `return_all_scores` is now deprecated,  if want a similar functionality use `top_k=None` instead of `return_all_scores=True` or `top_k=1` instead of `return_all_scores=False`.
  warnings.warn(
감정 분석 중:   0%|          | 0/4180 [00:00<?, ?it/s]WARNING:root:Oh! You have mecab in your environment. Kss will take this as a backend! :D

감정 분석 중: 100%|██████████| 4180/4180 [03:20<00:00, 20.81it/s]

분석 결과가 저장되었습니다: /content/drive/MyDrive/diary_sentiment_analysis/train_df_emotions_1.csv


#전체 코드 단계별 로직
##1. 데이터 및 모델 준비
CSV 파일에서 일기 텍스트(text 컬럼) 불러오기

KOTE 멀티라벨 감정 분류 모델 및 토크나이저 로드

KOTE 원본 43개 감정 → 13개 감정(형용사형+중립)으로 매핑 테이블 정의

##2. 텍스트 전처리
KSS 문장 분리기로 텍스트를 자연스러운 문장 단위로 분리

각 문장의 토큰 길이를 계산 (tokenizer.encode)

512 토큰 이하(특수 토큰 포함)로 여러 문장을 묶어 chunk 생성

한 문장이 너무 길면 단어 단위로 강제 분할하여 chunk 생성

##3. 감정 예측 및 후처리
각 chunk별로 감정 예측 (멀티라벨, sigmoid 확률)

모델이 반환한 원본 감정명(label)을 kote_to_13 매핑을 통해 13개 감정군으로 변환

threshold(예: 0.4) 이상 감정만 "존재"로 인정, 상위 top_n(예: 3개)만 추출

chunk별로 감정별 최대 확률 집계

##4. 최종 감정 결정
모든 감정 확률이 threshold 미만이면 '그저 그런'만 남김

아니면, 존재하는 감정 중 final_threshold(예: 0.6) 이상 감정만 최종 감정으로 인정 (내림차순 정렬 후 상위 3개)

최종 감정이 없다면 '그저 그런'만 남김

##5. 결과 저장
감정별 확률 딕셔너리(emotions_prob)와 최종 감정 리스트(exist_emotions)를 새 컬럼에 저장

결과를 새로운 CSV 파일로 저장


#### 저장 형태 변환 과정

In [ ]:
# 감정 저장 리스트 형태로 반환

import pandas as pd

data = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_1_emotions.csv")

# 1. 문자열을 리스트로 변환 (쉼표 기준 split, 공백 제거)
data['exist_emotions'] = data['exist_emotions'].apply(lambda x: [emotion.strip() for emotion in x.split(',') if emotion.strip()])

# 변환 결과 확인
data['exist_emotions'].head()

data.to_csv('/content/drive/MyDrive/diary_sentiment_analysis/train_batch_1_emotions.csv', index=False, encoding='utf-8-sig')

In [ ]:

import pandas as pd

data = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_2_emotions.csv")

# 1. 문자열을 리스트로 변환 (쉼표 기준 split, 공백 제거)
data['exist_emotions'] = data['exist_emotions'].apply(lambda x: [emotion.strip() for emotion in x.split(',') if emotion.strip()])

# 변환 결과 확인
data['exist_emotions'].head()

data.to_csv('/content/drive/MyDrive/diary_sentiment_analysis/train_batch_2_emotions.csv', index=False, encoding='utf-8-sig')

In [ ]:

import pandas as pd

data = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_3_emotions.csv")

# 1. 문자열을 리스트로 변환 (쉼표 기준 split, 공백 제거)
data['exist_emotions'] = data['exist_emotions'].apply(lambda x: [emotion.strip() for emotion in x.split(',') if emotion.strip()])

# 변환 결과 확인
data['exist_emotions'].head()


data.to_csv('/content/drive/MyDrive/diary_sentiment_analysis/train_batch_3_emotions.csv', index=False, encoding='utf-8-sig')

## 사전학습 모델 사용 vs llm(gemini) 사용 비교

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# 사전 학습 모델 사용
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_df_emotions_1.csv")
df

category              title  \
0     비출판물 > 일기    5학년을 마치며....      
1     비출판물 > 일기     <name1>의 깜짝 등장   
2     비출판물 > 일기       <name2>집에 초대   
3     비출판물 > 일기              Happy   
4     비출판물 > 일기                 TV   
...         ...                ...   
4175  비출판물 > 일기                힘들다   
4176  비출판물 > 일기         힘들면서 좋았던 날   
4177  비출판물 > 일기            힘들었던 날.   
4178  비출판물 > 일기       힘은 솟아나지 않는다.   
4179  비출판물 > 일기        힘을 내라고 말해줄래   

                                                   text  \
0     5학년을 마치며....                                  ...   
1     2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...   
2     <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...   
3     다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...   
4     앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...   
...                                                 ...   
4175  요즈음 엄마가 회사 일 때문에 바쁘셔서 밥을 못 챙겨 주신다. 나는 밥과 반찬을 잘...   
4176  <힘들면서 좋았던 날> 오늘 아침 8시에 일어났다. 조금 피곤했지만 또 잠들기는 싫...   
4177  오늘은 감정적으로 매우 힘든 날이었다. 우울하고 지치고... 나갈 곳은 없고 하고싶...   
4178  힘은 솟아나지 않는다. 힘이란 연속되면서 이어진다. 관성과 같다. 나는 언제나 솟구...   
4179  같이 게임하는 친구가 지칠때마다 힘들어도.. 힘내자! 이러는데 그 말이 너무 웃김 ...   

                                          emotions_prob  \
0     {'기쁜': 0.9251657128334045, '감사한': 0.8734253644...   
1     {'기쁜': 0.9728065133094788, '감사한': 0.4869050085...   
2     {'기쁜': 0.9636561870574951, '감사한': 0.6682934165...   
3     {'기쁜': 0.9263925552368164, '감사한': 0.7673888206...   
4     {'흥분한': 0.9628071188926697, '편안한': 0.650756776...   
...                                                 ...   
4175  {'슬픈': 0.7687985897064209, '스트레스 받는': 0.724810...   
4176  {'스트레스 받는': 0.7236539125442505, '기쁜': 0.625695...   
4177  {'슬픈': 0.9564365744590759, '스트레스 받는': 0.903291...   
4178  {'만족스러운': 0.8616605997085571, '슬픈': 0.88051468...   
4179  {'스트레스 받는': 0.892622172832489, '슬픈': 0.8368541...   

                exist_emotions  
0         ['기쁜', '감사한', '편안한']  
1                       ['기쁜']  
2                ['기쁜', '감사한']  
3                ['기쁜', '감사한']  
4        ['흥분한', '편안한', '감사한']  
...                        ...  
4175  ['슬픈', '스트레스 받는', '흥분한']  
4176         ['스트레스 받는', '기쁜']  
4177  ['슬픈', '스트레스 받는', '우울한']  
4178    ['슬픈', '만족스러운', '우울한']  
4179  ['스트레스 받는', '슬픈', '불안한']  

[4180 rows x 5 columns]

In [ ]:
# 상위 500개 데이터만 추출
pre_top500 = df.head(500)
pre_top500

category              title  \
0    비출판물 > 일기    5학년을 마치며....      
1    비출판물 > 일기     <name1>의 깜짝 등장   
2    비출판물 > 일기       <name2>집에 초대   
3    비출판물 > 일기              Happy   
4    비출판물 > 일기                 TV   
..         ...                ...   
495  비출판물 > 일기             가우디 투어   
496  비출판물 > 일기                 가을   
497  비출판물 > 일기             가을 산책    
498  비출판물 > 일기           가을 열매 관찰   
499  비출판물 > 일기            가을방학이오면   

                                                  text  \
0    5학년을 마치며....                                  ...   
1    2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...   
2    <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...   
3    다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...   
4    앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...   
..                                                 ...   
495  제목 : 가우 디투어(스페인) □ 새벽 5시에 일어나서 빨리 준비해 전기 택시를 타...   
496  가을은 천고마비의 계절이라고 한다. 하늘은 높고 말은 살찐다는 뜻이다. 형이 천고동...   
497  오늘우리가족 모두 뒷산에 올라 갔다.올라가면서 산수유,씨았,도토리,밤등을 주었다. ...   
498  2019년 10월 30일 수 날씨:눈 오는 날 처럼 추움 제목:가을 열매 관찰 오늘...   
499  2015년 9월 17일 木요일 날씨: 맑음 제목: 가을방학이 오면 가을 방학이 온다...   

                                         emotions_prob         exist_emotions  
0    {'기쁜': 0.9251657128334045, '감사한': 0.8734253644...   ['기쁜', '감사한', '편안한']  
1    {'기쁜': 0.9728065133094788, '감사한': 0.4869050085...                 ['기쁜']  
2    {'기쁜': 0.9636561870574951, '감사한': 0.6682934165...          ['기쁜', '감사한']  
3    {'기쁜': 0.9263925552368164, '감사한': 0.7673888206...          ['기쁜', '감사한']  
4    {'흥분한': 0.9628071188926697, '편안한': 0.650756776...  ['흥분한', '편안한', '감사한']  
..                                                 ...                    ...  
495  {'감사한': 0.9743372201919556, '기쁜': 0.8448079824...   ['감사한', '기쁜', '편안한']  
496  {'슬픈': 0.9309316873550415, '당황한': 0.7861250042...          ['슬픈', '당황한']  
497  {'기쁜': 0.9642611145973206, '감사한': 0.8898417949...   ['기쁜', '감사한', '흥분한']  
498  {'기쁜': 0.9626606702804565, '감사한': 0.9495124220...   ['기쁜', '감사한', '흥분한']  
499  {'기쁜': 0.6289383769035339, '흥분한': 0.5732958316...                 ['기쁜']  

[500 rows x 5 columns]

In [ ]:
# 사전 학습 모델 사용
import pandas as pd

gemini_df = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_1_emotions_no_lonely.csv")
gemini_df

category              title  \
0    비출판물 > 일기    5학년을 마치며....      
1    비출판물 > 일기     <name1>의 깜짝 등장   
2    비출판물 > 일기       <name2>집에 초대   
3    비출판물 > 일기              Happy   
4    비출판물 > 일기                 TV   
..         ...                ...   
495  비출판물 > 일기             가우디 투어   
496  비출판물 > 일기                 가을   
497  비출판물 > 일기             가을 산책    
498  비출판물 > 일기           가을 열매 관찰   
499  비출판물 > 일기            가을방학이오면   

                                                  text          exist_emotions  
0    5학년을 마치며....                                  ...     ['기쁜', '슬픈', '불안한']  
1    2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...  ['기쁜', '만족스러운', '흥분한']  
2    <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...         ['기쁜', '만족스러운']  
3    다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...  ['기쁜', '만족스러운', '감사한']  
4    앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...               ['그저 그런']  
..                                                 ...                     ...  
495  제목 : 가우 디투어(스페인) □ 새벽 5시에 일어나서 빨리 준비해 전기 택시를 타...  ['기쁜', '감사한', '만족스러운']  
496  가을은 천고마비의 계절이라고 한다. 하늘은 높고 말은 살찐다는 뜻이다. 형이 천고동...         ['슬픈', '그저 그런']  
497  오늘우리가족 모두 뒷산에 올라 갔다.올라가면서 산수유,씨았,도토리,밤등을 주었다. ...  ['만족스러운', '기쁜', '편안한']  
498  2019년 10월 30일 수 날씨:눈 오는 날 처럼 추움 제목:가을 열매 관찰 오늘...  ['흥분한', '기쁜', '만족스러운']  
499  2015년 9월 17일 木요일 날씨: 맑음 제목: 가을방학이 오면 가을 방학이 온다...                  ['기쁜']  

[500 rows x 4 columns]

In [ ]:
# Jaccard 유사도(다중 감정 일치도) 계산

from typing import List

def jaccard_similarity(list1: List[str], list2: List[str]) -> float:
    set1, set2 = set(list1), set(list2)
    if not set1 and not set2:
        return 1.0
    return len(set1 & set2) / len(set1 | set2)

# 각 행마다 Jaccard 유사도 계산
jaccard_scores = [
    jaccard_similarity(a, b)
    for a, b in zip(pre_top500['exist_emotions'], gemini_df['exist_emotions'])
]

# 평균 유사도 계산
mean_jaccard = sum(jaccard_scores) / len(jaccard_scores)
print(f"평균 Jaccard 유사도: {mean_jaccard:.3f}")


평균 Jaccard 유사도: 0.526


In [ ]:
# 감정별 분포 비교

import ast
from collections import Counter
import pandas as pd

# 문자열을 리스트로 변환
pre_top500['exist_emotions'] = pre_top500['exist_emotions'].apply(ast.literal_eval)
gemini_df['exist_emotions'] = gemini_df['exist_emotions'].apply(ast.literal_eval)

# 평탄화 및 집계
all_emotions_pre_top500 = [emotion for emotions in pre_top500['exist_emotions'] for emotion in emotions]
emotion_counts_pre_top500 = Counter(all_emotions_pre_top500)
result_pre_top500 = pd.DataFrame(emotion_counts_pre_top500.items(), columns=['emotion', 'count']).sort_values('count', ascending=False)

all_emotions_gemini_df = [emotion for emotions in gemini_df['exist_emotions'] for emotion in emotions]
emotion_counts_gemini_df = Counter(all_emotions_gemini_df)
result_gemini_df = pd.DataFrame(emotion_counts_gemini_df.items(), columns=['emotion', 'count']).sort_values('count', ascending=False)

print("pre_top500 감정별 분포")
print(result_pre_top500)

print("\ngemini_df 감정별 분포")
print(result_gemini_df)



pre_top500 감정별 분포
    emotion  count
0        기쁜    265
1       감사한    194
5        슬픈    165
4     만족스러운    143
3       흥분한    125
8   스트레스 받는    119
2       편안한    111
6       당황한     60
9       불안한     51
10      우울한     45
7       분노한     35
11    그저 그런      7
12     자신하는      5

gemini_df 감정별 분포
    emotion  count
3     만족스러운    224
0        기쁜    208
1        슬픈    198
6     그저 그런    160
2       불안한    126
5       감사한    113
11      우울한     78
4       흥분한     55
8       분노한     42
10  스트레스 받는     40
9       당황한     36
12      편안한     28
7      자신하는     26


<ipython-input-6-1334189653>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pre_top500['exist_emotions'] = pre_top500['exist_emotions'].apply(ast.literal_eval)


## 두 방식 비교

### 평균 Jaccard 유사도: 0.526

해석:
완전히 일치하지는 않지만, 절반 정도는 감정군이 겹친다는 뜻입니다.
다중 감정 분석에서 이 정도 유사도는, 감정 분류 체계가 동일하더라도 모델별 감정 해석 방식이 상당히 다름을 보여줍니다.


### 감정별 분포 차이

- 사전학습 모델이 더 많이 잡은 감정:
편안한, 감사한, 스트레스 받는, 흥분한, 기쁜, 당황한  

- LLM 모델이 더 많이 잡은 감정:
그저 그런, 만족스러운, 불안한, 우울한, 슬픈, 자신하는, 분노한  

- 특이점:  

        '그저 그런': LLM 모델은 160건, 사전학습 모델은 7건 → LLM 모델이 중립적 감정을 훨씬 더 많이 매핑

        '만족스러운': LLM 모델이 더 많이 잡음

        '편안한', '감사한', '스트레스 받는', '흥분한': 사전학습 모델이 상대적으로 더 많이 매핑


#### 해석 및 인사이트
- 모델별 감정 해석 경향 차이

  LLM 모델은 중립적이거나 복합적인 상황에 '그저 그런', '만족스러운', '불안한', '우울한', '슬픈' 등 감정을 더 적극적으로 부여

  사전학습 모델은 '기쁜', '감사한', '편안한', '스트레스 받는', '흥분한' 등 긍정적·활동적 감정을 더 자주 매핑

- 중립/복합 감정 처리

  LLM 모델이 '그저 그런' 등 중립 감정을 더 많이 감지하는 것은, 감정의 미묘함이나 애매함을 더 포착하려는 경향으로 볼 수 있음

- 감정 분포의 다양성

  LLM 모델은 감정 분포가 더 넓고, 사전학습 모델은 특정 감정에 집중되는 경향

In [ ]:
# 감정별 대표 사례 분석
# "같은 일기에서 Hugging Face는 '기쁜'으로, Gemini는 '만족스러운'으로 감정을 매핑(분류)한 사례"

for idx, (hf_emotions, gm_emotions) in enumerate(zip(pre_top500['exist_emotions'], gemini_df['exist_emotions'])):
    if '기쁜' in hf_emotions and '만족스러운' in gm_emotions and '기쁜' not in gm_emotions:
        print(f"Index: {idx}")
        print("일기 내용:", pre_top500.iloc[idx]['text'])
        print("Hugging Face 감정:", hf_emotions)
        print("Gemini 감정:", gm_emotions)
        print("-" * 40)
        # 필요시 break로 일부만 출력


Index: 18
일기 내용: 2019.10.28. 편지콘서트 내가 이콘서트를보았을 때 느낀점은 곡중에는 운명이 가장 마음이들었다 왜냐면가장웅장해서인것 같다. 그리고 베토벤이 표현한 운명이 문을 두드리는 소리는 내가생각하기에는 인생이 험난해서 주저앉을때 뭔가교훈을 주기위해 나는소리 같다 그래서 웅장하고 크게 만든음악 갔다 나는그런 웅장한노래가 부드러운 노래나 시끄러 운노래보다훨씬좋다. 그리고 나는이콘서트를 본 것도기억이나지만 제일신기했던건 우리학교가 이콘 서트를 볼 수 있게 뽑인것이 정말 신기했다. 몇백개학교중에서 우리학교 가 뽑인것은 정말 행운(?)이다. 그런데 곡이너무많아서 머리가 혼란 스러웠다 만약 다음에또본다면 시간을 좀 줄이는것을추천할거같다. 우리반애들도 거의다 지치고 심심해했던거같다. 그래도 다른 콘서트보다 헐씬재밌고 상황을설명해주어서 재밌었다.
Hugging Face 감정: ['감사한', '기쁜', '흥분한']
Gemini 감정: ['그저 그런', '만족스러운']
----------------------------------------
Index: 33
일기 내용: 출발전에는 여행만 하고 물건 사지 않겠다고 했지만 프로폴리스는 몸속의 염증 치료 폴리코사놀은 혈관 건강에 좋타고 해서 아무리돈이 없어도 살면서 고통없이 살려면  이건 대출을 내서라도 먹어야 되겠다 싶어 망설이지 않고 계약을했다 평소 손발이 시리고 저리고 뻣뻣하며 쥐가나면 통증이 심하고 남 아프다 사놓고 속은건 아닌지 후회도 했는데 꾸준히 1개월 정도 먹으니 효과 있었다 요즘은 종아리가 저리지도 쥐도 나지않고 통증이 거의 사라졌다 더 욕심을 내자면 무릅 관절염도 효과 봤으면 좋겠다~~ㅎ 관절염 3기초라는데 무릅 통증만 없으면 괜찷을거같다 매일 스커트 운동으로 관리중~~ 건강해야 삶의질도 높은것 약값이 좀 비싸지만 아깝다는 생각이 들지 않고 잘삿다 슆다 날마다 열심히 빠트 리지 않고 꼭꼭 잘 챙겨 먹어야지~~^^ 형편이 된다면 가족이 다같이  먹으면 좋겠다 애들꺼까지 사고 싶었는데 아쉽다~

In [ ]:
# 슬픈과 우울한 감정이 매핑된 일기 샘플 추출
for idx, (hf_emotions, gm_emotions) in enumerate(zip(pre_top500['exist_emotions'], gemini_df['exist_emotions'])):
    # 한 모델은 '슬픈', 다른 모델은 '우울한'만 있는 경우
    # Hugging Face와 Gemini 모두 '슬픈', '우울한'을 포함하거나, 한 모델은 '슬픈', 다른 모델은 '우울한'을 포함한 경우

    if ('슬픈' in hf_emotions and '우울한' in gm_emotions) or ('우울한' in hf_emotions and '슬픈' in gm_emotions):
        print(f"Index: {idx}")
        print("일기 내용:", pre_top500.iloc[idx]['text'])
        print("Hugging Face 감정:", hf_emotions)
        print("Gemini 감정:", gm_emotions)
        print("-" * 40)


Index: 20
일기 내용: 할머니. 내게 그리운 호칭, 그리운 사람, 그리운 감정, 한 평생에 느껴야 할 애상이 녹아내려 빚은 단어는 거창하지 않았다. 하얗게 센 머리와 세월을 이기지 못한 척추와 피부. 누군가는 피할지도 모르는 울퉁불퉁한 발걸음과 유독 동그랗던 이마, 웃음기도 없고 살가운 인상은 아닌 늙은 얼굴이 나에게 미소 지었던 순간을 떠올려본다. 그리 오래된 순간이 아니었음에도 나는 아득함을 느꼈다. 내 머리통의 크기가 당신을 모두 추억하기에 터무니없이 작아 서럽다. 내 머리통이 지구처럼 거대했다면 할머니께서 주신 모든 걸 다 기억할 수 있었을 텐데, 정작 내 비통함은 한숨에도 모조리 담기지 않을 만큼 가벼워서 가슴이 싸해졌다. 누구에게나 할머니는 존재한다. 그에 대한 마음은 천차만별일 지라도. 내게 할머니는 커다란 둥지였고, 늙은 어머니였고, 내 선생님이었고, 내 친구였고, 내 인생을 손으로 더듬어 쓰다듬어주었던 분이셨다. 내가 너무 당신을 거대하게 만들었다고 생각할 지도 모르겠지만, 안타깝게도 그 분께서는 내 옆에 있으셨을 적부터 눈이 나빠 큰아들이 보낸 편지를 내가 낭랑히 읽어드려야 했고, 내가 이 글을 쓸 때는 이미 할머니께서 내 곁을 떠난지 1년이 되는 날을 하루 앞두고 쓰는 글인지라 어렸을 때처럼 맑은 목소리로 낭랑하게 이 글을 모두 읽어도 나를 탓하실 수는 없으시다. 편지 내용을 낭랑한 손녀의 목소리로 조용히 들으시던 당신 눈동자는 계곡의 돌멩이를 닮았었다. 그 색이 청동을 닮아 내 얼굴이 코앞에 있지 않으면 보지 못하신다는 걸 알면서도 그 눈을 사랑했다. 나는 올해로 열아홉 살이다. 할머니는 작년에 돌아가셨다. 2019년 4월 12일. 내 수학여행은 11일에 끝났다. 비행기가 지연돼서 공항 바닥에서 노숙하는 줄만 알고 있다가 쓰러지기 직전에 내 방 침대에 누웠고 기절하듯이 잤다. 다음 날 아침 9시에 내 눈을 뜨게 만든 건 아빠의 목소리였다. 병원에 가야 돼. 할머니께서 돌아가셨대. 검은 옷으로 갈아입어. 내 인생의 기저가 푹 꺼졌는데

In [ ]:
# Gemini에서 '그저 그런'이 매핑된 사례 추출
for idx, (hf_emotions, gm_emotions) in enumerate(zip(pre_top500['exist_emotions'], gemini_df['exist_emotions'])):
    if '그저 그런' in gm_emotions:
        print(f"Index: {idx}")
        print("일기 내용:", pre_top500.iloc[idx]['text'])
        print("Hugging Face 감정:", hf_emotions)
        print("Gemini 감정:", gm_emotions)
        print("-" * 40)

Index: 4
일기 내용: 앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아해서 머리가 나빠질수도 있으니까 TV를 봐도 머리가 안나빠지고 좋아지는 TV가 꼭 발명 되어야 하는 발명품입니다. (나 때문에)
Hugging Face 감정: ['흥분한', '편안한', '감사한']
Gemini 감정: ['그저 그런']
----------------------------------------
Index: 16
일기 내용: 코로나 때문에 학원도 쉬고 학교 개학도 일주일 더 연기됐다 이번에 학원을 한 달 동안 끊고 인강을 들으면서 문제집을 풀기로 했는데 무슨 인강이 좋은지 무슨 학습지가 좋은지 아직 알아봐야 할 것 같다 며칠 동안 안 나가고 집에 있으니 심심하다 할 것도 없다 평소에 가기 싫던 학교도 가고 싶다 그리고 일주일? 동안 애기들이 왔다 갔는데 내 물건 만지지 마라 시끄럽다 물은 니가 알아서 떠다 먹어라 소리 지르지 마라 귀청 떨어진다 나도 어렸을 때 저랬을 거라고 하지만 그렇다고 해서 내가 꼭 놀아줘야 하나... 너네한테 관심 주기 싫다 너네랑 놀기 싫다 나 따라 하지 마라 뭐가 이쁜지 모르겠다
Hugging Face 감정: ['스트레스 받는', '분노한']
Gemini 감정: ['분노한', '그저 그런', '슬픈']
----------------------------------------
Index: 17
일기 내용: < 통화 > 선생님과 통화를 했다, 첫통화였다. 근데 4월인데 선생님을 한번도 못봤다. 한번 보긴 했지, 사진으로 ... 근데 선생님이 사진보다 안예쁘다고 했다. 그 말이 재밌었다. 코로나를 빨리 이겨서 학교에 가야겠다. 그러기 위해서 난 내일도 집콕방콕
Hugging Face 감정: ['스트레스 받는', '불안한']
Gemini 감정: ['그저 그런']
----------------------------------------
Index: 18
일기 내용: 2019.10.28. 편지콘서트 내가 이콘서트를보았을 때

#### 1. 모델별 감정 해석 경향  
Hugging Face 모델  

즉각적이고 명확한 긍정(‘기쁜’, ‘감사한’, ‘흥분한’ 등)이나 부정(‘슬픈’, ‘스트레스 받는’ 등) 감정을 주로 매핑.

감정의 복합성이나 중립성(‘그저 그런’)보다는 한두 가지 뚜렷한 감정에 집중하는 경향이 있음.

감정의 미묘한 경계나 애매함보다는, 분명한 감정 라벨을 선호하는 결과가 많음.

Gemini 모델  

감정의 복합성, 중립성, 애매함(‘그저 그런’, ‘만족스러운’, ‘불안한’, ‘우울한’ 등)을 더 섬세하게 포착함.

한 일기에 여러 감정이 혼재된 경우, 그 층위와 결까지 반영하려는 경향이 뚜렷함.

특히 ‘그저 그런’ 등 중립적 감정이나, ‘만족스러운’, ‘우울한’, ‘불안한’ 등 미묘한 감정의 구분이 잘 드러남.  

#### 2. 감정별 분포 및 일치도  
두 모델의 감정 분포는 상위 감정군에서는 겹치지만, ‘그저 그런’ 등 중립적 감정이나 ‘만족스러운’, ‘우울한’ 등 복합 감정에서 차이가 큼.

평균 Jaccard 유사도(0.526)는 두 모델이 절반 정도만 감정 조합에서 일치한다는 의미로, 감정 해석 방식이 상당히 다름을 보여줌.

Gemini는 감정의 애매함, 복합성, 무감정 상태를 더 자주 매핑하는 반면, Hugging Face는 보다 단순·직설적인 감정 분류에 집중함.

#### 3. 대표 사례 분석  
긍정 감정:  
Hugging Face는 ‘기쁜’ 등 즉각적 긍정에 집중, Gemini는 ‘만족스러운’, ‘감사한’ 등 여운과 복합적 긍정 감정을 더 자주 포착.

부정 감정:  
‘슬픈’과 ‘우울한’의 경계에서 Gemini는 맥락에 따라 더 세밀하게 구분, Hugging Face는 ‘슬픈’을 기본값처럼 넓게 적용.

중립/복합 감정:  
Gemini는 ‘그저 그런’ 등 중립적 감정을 적극적으로 매핑, Hugging Face는 거의 매핑하지 않음.

#### 4. 실제 일기 맥락과의 적합성  
 감정이 복합적이거나 애매한 일기에서는 Gemini의 결과가 더 현실적이고 섬세하게 느껴지는 경우가 많았음.  

 감정을 명확히 구분하거나, 빠르게 분류해야 하는 상황에서는 Hugging Face 결과가 더 단순·직관적일 수 있음.

#### 5. 객관적 결론  
감정의 복합성, 미묘함, 중립성까지 포착하고 싶다면 Gemini 모델이 더 적합함.

명확하고 단순한 감정 분류가 목적이라면 Hugging Face 모델이 유리함.

## LLM 모델(Gemini) 사용: '외로운' 감정을 추가해야 할지 안해도 될지

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

df_lonely = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_1_emotions.csv")
df_no_lonely = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_1_emotions_no_lonely.csv")

In [ ]:
import ast

# exist_emotions가 문자열이면 리스트로 변환
df_lonely['exist_emotions'] = df_lonely['exist_emotions'].apply(ast.literal_eval)
lonely_only = df_lonely[df_lonely['exist_emotions'].apply(lambda emotions: '외로운' in emotions)]

In [ ]:
# 인덱스 기준으로 비교 (lonely_only의 index와 df_no_lonely의 index가 동일하다고 가정)
for idx in lonely_only.index:
    print(f"Index: {idx}")
    print("일기 내용:", lonely_only.loc[idx, 'text'])
    print("외로운 포함 감정:", lonely_only.loc[idx, 'exist_emotions'])
    print("외로운 제외 감정:", ast.literal_eval(df_no_lonely.loc[idx, 'exist_emotions']))
    print("-" * 40)

Index: 44
일기 내용: 아이디랑 비밀번호 통일하는 일은 왜 아렇게 어렵지. 한 사람인데도 이렇게 의견 통일 안되는데 다른 사람하고 의견합 맞추는 게 어디 쉽겠냐 환자분 성함이 <name1>여서 톰과제리인가 ㅋㅋ 유치한 생각하고 있었는데 신고 있던 양말이 진짜 재리 였음 다른 사람에게 사랑받는 거 좋은데 일단 나는 내가 사랑할 수 없다는 것에서 엄청난 공허함을 느낀다 장르내놔.
외로운 포함 감정: ['슬픈', '외로운', '그저 그런']
외로운 제외 감정: ['슬픈', '그저 그런', '불안한']
----------------------------------------
Index: 47
일기 내용: 10시, 늦은 밤은 아니지만 겨울의 10시는 춥고 어둡다. 나름 시내라고 불리는 역에서 천천히 걸어 집에 가는 길이다. 12월 중순의 사람들은 귀가 빨개져 춥다며 그 유행 중인 롱패딩의 지퍼를 코 밑까지 올렸다. 웃기게도 난 야상을 벗어든 채 걸었다. "덥다"라는 말은 입에서 나옴과 동시에 입김과 함께 금방 흩어져버렸다. 차도 지나다니는 꽤 넓은 길인데 사람이 몇 없다. 춥다고 다 사라진 것일까. 가끔 이른 밤길을 혼자 걸을 떈 이상한 기분이 들기도 한다. 세상에 혼자 있는 느낌. 입으로 반복적인 입김을 내뱉어보니 내가 주전자가 된 느낌이다. 집에 가까워졌다. 좁은 길에 들어선다. 나름 가로등이 6개나 있는 골목이지만 망가져서 불빛이 드는 건 2개 뿐이다. 시선이 입김을 따라 하늘로 갔다. 밤하늘을 정말 좋아하지만 여긴 귀신같은 전기줄이 추욱 쳐져있어 으스스한 분위기를 내뿜는 것과 동시에 밤하늘을 가리고 있다. 그래도 달은 빛난다. 어쩐지 초라해진 느낌이다. 세상이 어떻건 전깃줄이 하늘을 가려도, 날씨가 추워도, 아무도 하늘을 보지 않아도 달은 차가운 빛을 쏟고 있다. 나는 그 빛에 베어 아무것도 할 수 없다. 혹시 나도 빛을 낼 수 있을까 작은 별이 될 수 있을까 눈 앞에 있는 전기줄을 모두 잘라 없애고 싶다. 그럼 작은 별빛도 볼 수 있을텐데
외로운 포함 

‘외로운’이 빠진 경우,  

‘슬픈’, ‘우울한’, ‘불안한’, ‘그저 그런’ 등으로만 매핑되는데,

이 감정들은 관계적 고립감, 소속 결여, 실존적 허전함을 충분히 포착하지 못합니다.

“외로운”이 포함된 버전의 감정 매핑이 일기 내용의 핵심 정서(고독, 소외, 허전함,  관계 결핍 등)를 더 정확하게 반영

=> '외로운' 감정은 필요한 감정임.

## LLM 모델(Gemini)을 사용한 감정 매핑 결과 통합하기

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

df1 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_1_emotions.csv")
df2 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_2_emotions.csv")
df3 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_3_emotions.csv")
df4 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_4_emotions.csv")
df5 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_5_emotions.csv")
df6 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_6_emotions.csv")
df7 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_7_emotions.csv")
df8 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_8_emotions.csv")
df9 = pd.read_csv("/content/drive/MyDrive/diary_sentiment_analysis/train_batch_9_emotions.csv")

In [ ]:
print(df1.head())
print(df2.head())
print(df3.head())
print(df4.head())
print(df5.head())
print(df6.head())
print(df7.head())
print(df8.head())
print(df9.head())

    category              title  \
0  비출판물 > 일기    5학년을 마치며....      
1  비출판물 > 일기     <name1>의 깜짝 등장   
2  비출판물 > 일기       <name2>집에 초대   
3  비출판물 > 일기              Happy   
4  비출판물 > 일기                 TV   

                                                text          exist_emotions  
0  5학년을 마치며....                                  ...     ['기쁜', '감사한', '슬픈']  
1  2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...         ['기쁜', '만족스러운']  
2  <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...         ['기쁜', '만족스러운']  
3  다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...  ['기쁜', '만족스러운', '감사한']  
4  앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...         ['기쁜', '만족스러운']  
    category                           title  \
0  비출판물 > 일기                         가을의 일기1   
1  비출판물 > 일기                         가을저녁 추석   
2  비출판물 > 일기  가장 친한 친구의 첫번째 브라이덜샤워, 첫번째 결혼식!   
3  비출판물 > 일기                              가족   
4  비출판물 > 일기                           가족 여행   

                          

In [ ]:
dfs = [df1, df2, df3, df4, df5, df6, df7, df8, df9]
df = pd.concat(dfs, ignore_index=True)
df

category              title  \
0     비출판물 > 일기    5학년을 마치며....      
1     비출판물 > 일기     <name1>의 깜짝 등장   
2     비출판물 > 일기       <name2>집에 초대   
3     비출판물 > 일기              Happy   
4     비출판물 > 일기                 TV   
...         ...                ...   
4175  비출판물 > 일기                힘들다   
4176  비출판물 > 일기         힘들면서 좋았던 날   
4177  비출판물 > 일기            힘들었던 날.   
4178  비출판물 > 일기       힘은 솟아나지 않는다.   
4179  비출판물 > 일기        힘을 내라고 말해줄래   

                                                   text  \
0     5학년을 마치며....                                  ...   
1     2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...   
2     <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...   
3     다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...   
4     앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...   
...                                                 ...   
4175  요즈음 엄마가 회사 일 때문에 바쁘셔서 밥을 못 챙겨 주신다. 나는 밥과 반찬을 잘...   
4176  <힘들면서 좋았던 날> 오늘 아침 8시에 일어났다. 조금 피곤했지만 또 잠들기는 싫...   
4177  오늘은 감정적으로 매우 힘든 날이었다. 우울하고 지치고... 나갈 곳은 없고 하고싶...   
4178  힘은 솟아나지 않는다. 힘이란 연속되면서 이어진다. 관성과 같다. 나는 언제나 솟구...   
4179  같이 게임하는 친구가 지칠때마다 힘들어도.. 힘내자! 이러는데 그 말이 너무 웃김 ...   

                exist_emotions  
0          ['기쁜', '감사한', '슬픈']  
1              ['기쁜', '만족스러운']  
2              ['기쁜', '만족스러운']  
3       ['기쁜', '만족스러운', '감사한']  
4              ['기쁜', '만족스러운']  
...                        ...  
4175      ['슬픈', '흥분한', '불안한']  
4176                 ['그저 그런']  
4177      ['우울한', '슬픈', '외로운']  
4178      ['우울한', '슬픈', '외로운']  
4179  ['슬픈', '스트레스 받는', '우울한']  

[4180 rows x 4 columns]

In [ ]:
df.isnull().sum()

category          0
title             0
text              0
exist_emotions    0
dtype: int64

In [ ]:
from collections import Counter
import pandas as pd
import ast

# 안전하게 리스트로 변환
def safe_literal_eval(val):
    if isinstance(val, str):
        return ast.literal_eval(val)
    return val

df['exist_emotions'] = df['exist_emotions'].apply(safe_literal_eval)

# 평탄화 및 집계
all_emotions_df = [emotion for emotions in df['exist_emotions'] for emotion in emotions]
emotion_counts_df = Counter(all_emotions_df)
result_df = pd.DataFrame(emotion_counts_df.items(), columns=['emotion', 'count']).sort_values('count', ascending=False)

print("df 감정별 분포")
print(result_df)

df 감정별 분포
    emotion  count
3     만족스러운   2061
0        기쁜   2043
2        슬픈   1364
1       감사한    982
7     그저 그런    962
6       불안한    801
10      흥분한    675
11      우울한    447
8   스트레스 받는    414
5       분노한    394
9       당황한    307
4      자신하는    249
13      편안한    212
12      외로운    205


In [ ]:
df

category              title  \
0     비출판물 > 일기    5학년을 마치며....      
1     비출판물 > 일기     <name1>의 깜짝 등장   
2     비출판물 > 일기       <name2>집에 초대   
3     비출판물 > 일기              Happy   
4     비출판물 > 일기                 TV   
...         ...                ...   
4175  비출판물 > 일기                힘들다   
4176  비출판물 > 일기         힘들면서 좋았던 날   
4177  비출판물 > 일기            힘들었던 날.   
4178  비출판물 > 일기       힘은 솟아나지 않는다.   
4179  비출판물 > 일기        힘을 내라고 말해줄래   

                                                   text      exist_emotions  
0     5학년을 마치며....                                  ...       [기쁜, 감사한, 슬픈]  
1     2018/1/20 토요일 날씨 오늘은 <name2>이의 유치원친구들이 놀러왔다.일등...         [기쁜, 만족스러운]  
2     <name1>누나랑 같이 <name2>집에같다 <name2> 놀이터에서 물이나와 서...         [기쁜, 만족스러운]  
3     다언어 합창제를 했다. 우리 라보 노래 제목은 Happy였다. 그런데 내 짝꿍이 나...    [기쁜, 만족스러운, 감사한]  
4     앞으로 꼭 발명되어야 하든 발명품은 똑똑상자 TV입니다. 왜냐하면 제가 TV를 좋아...         [기쁜, 만족스러운]  
...                                                 ...                 ...  
4175  요즈음 엄마가 회사 일 때문에 바쁘셔서 밥을 못 챙겨 주신다. 나는 밥과 반찬을 잘...      [슬픈, 흥분한, 불안한]  
4176  <힘들면서 좋았던 날> 오늘 아침 8시에 일어났다. 조금 피곤했지만 또 잠들기는 싫...             [그저 그런]  
4177  오늘은 감정적으로 매우 힘든 날이었다. 우울하고 지치고... 나갈 곳은 없고 하고싶...      [우울한, 슬픈, 외로운]  
4178  힘은 솟아나지 않는다. 힘이란 연속되면서 이어진다. 관성과 같다. 나는 언제나 솟구...      [우울한, 슬픈, 외로운]  
4179  같이 게임하는 친구가 지칠때마다 힘들어도.. 힘내자! 이러는데 그 말이 너무 웃김 ...  [슬픈, 스트레스 받는, 우울한]  

[4180 rows x 4 columns]

In [ ]:
df.to_csv('/content/drive/MyDrive/diary_sentiment_analysis/train_diary_with_emotions.csv', index=False, encoding='utf-8-sig')